File: Algorithms/269. Alien Dictionary/AlienOrder.java
//Author: Tushar Jaiswal
//Creation Date: 11/28/2020

/*There is a new alien language that uses the English alphabet. However, the order among letters are unknown to you.

You are given a list of strings words from the dictionary, where words are sorted lexicographically by the rules of this new language.

Derive the order of letters in this language, and return it. If the given input is invalid, return "". If there are multiple valid solutions, return any of them.

Example 1:
Input: words = ["wrt","wrf","er","ett","rftt"]
Output: "wertf"

Example 2:
Input: words = ["z","x"]
Output: "zx"

Example 3:
Input: words = ["z","x","z"]
Output: ""
Explanation: The order is invalid, so return "".

Constraints:
    1 <= words.length <= 100
    1 <= words[i].length <= 100
    words[i] consists of only lowercase English letters.*/

/*Runtime Complexity: O(sum of lengths of all words)
Space Complexity: O(U + min(U^2,N)) where U is the total number of unique letters in the alien alphabet and N is the total number of strings in the input list.*/

class Solution {
    public String alienOrder(String[] words) {
        if (words.length == 1) {
            return words[0];
        }

        var orderMap = new HashMap<Character, HashSet<Character>>();
        HashMap<Character, Integer> inDegreeMap = createInDegreeMap(words);

        for (int i = 0; i < words.length - 1; i++) {
            if (!compareWords(words[i], words[i + 1], orderMap, inDegreeMap)) {
                return "";
            }
        }

        String result = topologicalSorting(orderMap, inDegreeMap);
        int visitedCharsCount = result.length();
        int totalCharsCount = inDegreeMap.size();

        return visitedCharsCount == totalCharsCount ? result : "";
    }

    private HashMap<Character, Integer> createInDegreeMap(String[] words) {
        HashMap<Character, Integer> inDegreeMap = new HashMap<Character, Integer>();
        for (String word: words) {
            for (char c : word.toCharArray())
            inDegreeMap.put(c, 0);
        }

        return inDegreeMap;
    }

    private boolean compareWords(String word1, String word2, HashMap<Character, HashSet<Character>> orderMap, HashMap<Character, Integer> inDegreeMap) {
        int len1 = word1.length();
        int len2 = word2.length();
        boolean noCharDiff = true;

        int i = 0;
        for (; i < Math.min(len1, len2); i++) {
            char c1 = word1.charAt(i);
            char c2 = word2.charAt(i);
            if (c1 != c2) {
                noCharDiff = false;
                if (!orderMap.containsKey(c1)) {
                    orderMap.put(c1, new HashSet<Character>());
                }
                if (!orderMap.get(c1).contains(c2)) {
                    orderMap.get(c1).add(c2);
                    inDegreeMap.merge(c2, 1, Integer::sum);
                }
                break;
            }
        }

        return len1 > len2 && noCharDiff ? false : true;
    }

    private String topologicalSorting(HashMap<Character, HashSet<Character>> orderMap, HashMap<Character, Integer> inDegreeMap) {
        var queue = new ArrayDeque<Character>();

        for (var entry : inDegreeMap.entrySet()) {
            if (entry.getValue() == 0) {
                queue.add(entry.getKey());
            }
        }

        StringBuilder result = new StringBuilder();
        while (!queue.isEmpty()) {
            char c = queue.remove();
            result.append(c);

            if (orderMap.containsKey(c)) {
                for (char child : orderMap.get(c)) {
                    inDegreeMap.merge(child, -1, Integer::sum);
                    if (inDegreeMap.get(child) == 0) {
                        queue.add(child);
                    }
                }
            }
        }

        return result.toString();
    }
}

File: Algorithms/269. Alien Dictionary/AlienOrder.cs
//Author: Tushar Jaiswal
//Creation Date: 06/23/2019

/*There is a new alien language which uses the latin alphabet. However, the order among letters are unknown to you. You receive a list of non-empty words from the dictionary, where words are sorted lexicographically by the rules of this new language. Derive the order of letters in this language.
Example 1:
Given the following words in dictionary,
[
  "wrt",
  "wrf",
  "er",
  "ett",
  "rftt"
]
The correct order is: "wertf".

Example 2:
Given the following words in dictionary,
[
  "z",
  "x"
]
The correct order is: "zx".

Example 3:
Given the following words in dictionary,
[
  "z",
  "x",
  "z"
]
The order is invalid, so return "".

Note:
You may assume all letters are in lowercase.
You may assume that if a is a prefix of b, then a must appear before b in the given dictionary.
If the order is invalid, return an empty string.
There may be multiple valid order of letters, return any one of them is fine.*/

using System;
using System.Collections.Generic;

namespace ConsoleApp
{
    class AlienDictionary
    {
        public static void Main(string[] args)
        {
            Console.WriteLine(AlienOrder(new string[] { "wrt", "wrf", "er", "ett", "rftt" }));
            Console.WriteLine(AlienOrder(new string[] { "z", "x" }));
            Console.WriteLine(AlienOrder(new string[] { "z", "x" , "z"}));
        }

        public static string AlienOrder(string[] words)
        {
            Dictionary<char, List<char>> graph = new Dictionary<char, List<char>>();
            Dictionary<char, int> inDegree = new Dictionary<char, int>();
            CreateGraph(words, graph, inDegree);
            string order = TopologicalSorting(graph, inDegree);
            return order.Length == graph.Count ? order : "";
        }

        private static void CreateGraph(string[] words, Dictionary<char, List<char>> graph, Dictionary<char, int> inDegree)
        {
            foreach (string word in words)
            {
                foreach (char c in word)
                {
                    if (!graph.ContainsKey(c))
                    { graph.Add(c, new List<char>()); }
                    if (!inDegree.ContainsKey(c))
                    { inDegree.Add(c, 0); }
                }
            }

            for (int i = 1; i < words.Length; i++)
            {
                string first = words[i - 1];
                string second = words[i];
                int minLength = Math.Min(first.Length, second.Length);

                for (int j = 0; j < minLength; j++)
                {
                    char parent = first[j];
                    char child = second[j];
                    if (parent != child)
                    {
                        graph[parent].Add(child);
                        inDegree[child]++;
                    }
                }
            }
        }

        private static string TopologicalSorting(Dictionary<char, List<char>> graph, Dictionary<char, int> inDegree)
        {
            Queue<char> queue = new Queue<char>();
            string order = "";

            foreach (KeyValuePair<char, int> kvp in inDegree)
            {
                if(kvp.Value == 0)
                { queue.Enqueue(kvp.Key); }
            }
            while(queue.Count != 0)
            {
                char c = queue.Dequeue();
                order += c;
                foreach(char child in graph[c])
                {
                    inDegree[child]--;
                    if(inDegree[child] == 0)
                    { queue.Enqueue(child); }
                }
            }
            return order;
        }
    }
}
